src/neo/Network/P2P/Peer.cs

```diff
@@ -174,29 +175,8 @@ private void OnStart(ChannelsConfig config)
             }
         }
 
-        private void OnTcpConnected(IPEndPoint remote, IPEndPoint local)
-        {
-            ImmutableInterlocked.Update(ref ConnectingPeers, p => p.Remove(remote));
-            if (MaxConnections != -1 && ConnectedPeers.Count >= MaxConnections && !TrustedIpAddresses.Contains(remote.Address))
-            {
-                Sender.Tell(Tcp.Abort.Instance);
-                return;
-            }
+        protected abstract void OnTcpConnected(IPEndPoint remote, IPEndPoint local);
```

Comment:
Nice abstraction, it looks like it may solve what @erikzhang suggested.

Reply:
I don't think moving this logic to subclass is good.

Reply:
It's a little difficult to completely decouple. Currently, I have keep the `OnTcpConnected` method,  but moved the connection validation to the subclass, as the disconnection reason was dependent on the validation logic. Plz have a review